### PR TITLE
pacemaker: added support for op defaults

### PR DESCRIPTION
**Pull request by AbelNavarro**
**Note:** this is a rework from https://github.com/crowbar/crowbar-ha/pull/262

We needed to add a default option for the pacemaker op defaults. In
particular we are interested in offering a default for "op monitor
on-fail" action.

Pacemaker does not offer a configuration option to offer such defaults
for all resources. There's a way to do it for a single resource with
rsc_defaults, but that is not our case. Also with Pacemaker you could
link one resource operations to another one with 'id-ref', but in that
case if you modify the reference resource it will cascade changes to all
other resourced.

This approach allows to store in the pacemaker barclamp a default for
op. In this case we are only interested in op monitor but could easily
be extended later on.

The primitive will read that value from the founder node and set it to
all 'op monitor on-fail' actions. Should the user had configured
manually the defaults (editing the barclamp), then those manually
configured values will prevail.

### CODE REVIEW

File: chef/cookbooks/pacemaker/providers/primitive.rb [hunk 1]
```diff
@@ -97,6 +104,25 @@ def maybe_modify_resource(name)
 
   cmds = []
 
+  ops = new_resource.op
+  ops = ops.to_hash unless ops.is_a? Hash
+  op_defaults = op_defaults(node)
+
+  # If op_defaults is defined and not nil, set the value
+  # otherwise remove it
+  if !op_defaults.nil? &&
```
Comment (aspiers) [inline chef/cookbooks/pacemaker/providers/primitive.rb]: The comment here is a code smell.  You can get rid of it by simplifying the huge conditional expression via "extract temporary variable" refactorings, and using clearly named temporary variables, e.g. (untested)

    on_fail_default_specified =
        !op_defaults.nil? &&
        op_defaults.key?("monitor") &&
        op_defaults["monitor"].key?("on-fail") &&
        !op_defaults["monitor"]["on-fail"].nil? &&
        !op_defaults["monitor"]["on-fail"].empty?

    monitor_op_needs_default =
        ops.key?("monitor") &&
        ops["monitor"].is_a?(Hash) &&
        !ops["monitor"].key?("on-fail")

    if on_fail_default_specified && monitor_op_needs_default
      ...
    end

Notice also how I avoided the need for parentheses like you had around this:

    (ops["monitor"].is_a? Hash)

Small cosmetic style tweak but I think my way is a bit more consistent with the other parts of the conditional, and hence very slightly easier on the eyes.

File: chef/cookbooks/pacemaker/providers/primitive.rb [hunk 2]
```diff
@@ -97,6 +104,25 @@ def maybe_modify_resource(name)
 
   cmds = []
 
+  ops = new_resource.op
+  ops = ops.to_hash unless ops.is_a? Hash
```
Comment (aspiers) [inline chef/cookbooks/pacemaker/providers/primitive.rb]: I think the motivation for ensuring it's a `Hash` (and thereby avoiding accidentally changing any C::N::A as a side-effect) is obscure enough that it warrants a comment explaining it.

Comment (AbelNavarro) [reply chef/cookbooks/pacemaker/providers/primitive.rb]: agreed. Comment will be included in the next commit update.

File: chef/cookbooks/pacemaker/providers/primitive.rb (hunk 1 above)
Comment (AbelNavarro) [reply chef/cookbooks/pacemaker/providers/primitive.rb]: Agree with the cosmetic changes. I have implemented those partially, mainly because the original motivation behind `(ops["monitor"].is_a? Hash)`. That exception was in place for the recent rabbitmq changes that allow two different monitor configurations per service, mainly to differentiate between master and slave roles. In that case, `monitor` is an array with the two configurations (Hash) instead of a plain Hash. The new code should deal with this situation.

**Comment by AbelNavarro** [comment]
[Rework on-fail for stop actions setting in pacemaker](https://trello.com/c/Jkr9NDFX/112-rework-on-fail-for-stop-actions-setting-in-pacemaker)

File: chef/cookbooks/pacemaker/providers/primitive.rb [hunk 3]
```diff
@@ -90,13 +97,48 @@ def update_resource(name)
...
+    !op_defaults.nil? &&
+    op_defaults.key?("monitor") &&
+    op_defaults["monitor"].key?("on-fail") &&
+    !op_defaults["monitor"]["on-fail"].nil? &&
+    !op_defaults["monitor"]["on-fail"].empty?
+
+  if on_fail_default_specified && ops.key?("monitor")
+    if ops["monitor"].is_a?(Array)
+      ops["monitor"].map! do |op_monitor|
+        op_monitor = set_op_default(op_monitor, op_defaults)
```
[the remark this answers is not in the document]
Comment (aspiers) [reply chef/cookbooks/pacemaker/providers/primitive.rb]: Please fix this warning, which is correct.

**Comment by AbelNavarro** [comment]
@aspiers fixed that map assignment.

File: chef/cookbooks/pacemaker/providers/primitive.rb [hunk 4]
```diff
@@ -69,6 +69,13 @@ def create_resource(name)
   standard_create_resource
 end
 
+def op_defaults(node)
+  return nil unless node[:pacemaker][:founder]
```
Comment (vuntz) [inline chef/cookbooks/pacemaker/providers/primitive.rb]: Why check for founder? I think it's actually wrong since it could lead to inconsistent results if something is run on founder and also on a non-founder node.

File: chef/data_bags/crowbar/template-pacemaker.json [hunk 5]
```diff
@@ -54,7 +54,12 @@
...
-      "clone_stateless_services": true
+      "clone_stateless_services": true,
+      "op_defaults": {
+        "monitor": {
+          "on-fail": ""
+        }
+      }
     }
   },
   "deployment": {
```
Comment (vuntz) [inline chef/data_bags/crowbar/template-pacemaker.json]: You need to bump the schema revision in this file (and adjust the migration filename). I guess there was a migration added since you initially started working on this.

File: chef/cookbooks/pacemaker/providers/primitive.rb [hunk 6]
```diff
@@ -90,13 +97,48 @@ def update_resource(name)
...
+
+  on_fail_default_specified =
+    !op_defaults.nil? &&
+    op_defaults.key?("monitor") &&
+    op_defaults["monitor"].key?("on-fail") &&
+    !op_defaults["monitor"]["on-fail"].nil? &&
+    !op_defaults["monitor"]["on-fail"].empty?
+
+  if on_fail_default_specified && ops.key?("monitor")
+    if ops["monitor"].is_a?(Array)
```
Comment (vuntz) [inline chef/cookbooks/pacemaker/providers/primitive.rb]: Wondering out loud: can it really be an `Array`? Do you have an example?

File: chef/cookbooks/pacemaker/providers/primitive.rb [hunk 7]
```diff
@@ -90,13 +97,48 @@ def update_resource(name)
...
+  op_defaults = op_defaults(node)
+
+  on_fail_default_specified =
+    !op_defaults.nil? &&
+    op_defaults.key?("monitor") &&
+    op_defaults["monitor"].key?("on-fail") &&
+    !op_defaults["monitor"]["on-fail"].nil? &&
+    !op_defaults["monitor"]["on-fail"].empty?
+
+  if on_fail_default_specified && ops.key?("monitor")
```
Comment (vuntz) [inline chef/cookbooks/pacemaker/providers/primitive.rb]: Don't we want to still set the default even if the primitive definition in the recipe doesn't contain anything about `"monitor"`?

File: chef/cookbooks/pacemaker/providers/primitive.rb [hunk 8]
```diff
@@ -90,13 +97,48 @@ def update_resource(name)
...
+  ops = ops.to_hash unless ops.is_a? Hash
+
+  op_defaults = op_defaults(node)
+
+  on_fail_default_specified =
+    !op_defaults.nil? &&
+    op_defaults.key?("monitor") &&
+    op_defaults["monitor"].key?("on-fail") &&
+    !op_defaults["monitor"]["on-fail"].nil? &&
+    !op_defaults["monitor"]["on-fail"].empty?
```
Comment (vuntz) [inline chef/cookbooks/pacemaker/providers/primitive.rb]: I somehow would move all this to `set_op_default`, see other comment.

File: chef/cookbooks/pacemaker/providers/primitive.rb [hunk 9]
```diff
@@ -90,13 +97,48 @@ def update_resource(name)
   maybe_modify_resource(name)
 end
 
+def set_op_default(op_monitor, op_defaults)
+  if !op_monitor.key?("on-fail")
+    op_monitor.merge("on-fail" => op_defaults["monitor"]["on-fail"])
+  else
+    op_monitor
+  end
```
Comment (vuntz) [inline chef/cookbooks/pacemaker/providers/primitive.rb]: First thing I wonder: wouldn't this work in all cases (so no need for if/else):

```ruby
default_op_monitor = { "on-fail" => op_defaults["monitor"]["on-fail"] }
default_op_monitor.merge(op_monitor)
```

Comment (vuntz) [inline chef/cookbooks/pacemaker/providers/primitive.rb]: Also, as said in a comment below, I'd move all the test on whether op_defaults contain something useful to this method.

So it'd be more like:

```ruby
# these three lines could be moved to the `op_defaults` method
op_defaults ||= {}
op_defaults["monitor"] ||= {}
op_defaults["monitor"]["on-fail"] ||= ""
if op_defaults.nil? || op_defaults["monitor"]["on-fail"].empty?
  op_monitor
else
  <current logic>
end
```

File: chef/cookbooks/pacemaker/providers/primitive.rb (hunk 4 above)
Comment (AbelNavarro) [reply chef/cookbooks/pacemaker/providers/primitive.rb]: Right, removing that. The reasoning for that is that we don't want the default to be set in different places. But this is already taken care of on the pacemaker_primitive invocation if needed: i.e.:

```
pacemaker_primitive smtp_resource do
  agent node[:pacemaker][:notifications][:agent]
  action [:stop, :delete]
  only_if "crm configure show #{smtp_resource}"
  only_if { CrowbarPacemakerHelper.is_cluster_founder?(node) }
end
```

File: chef/cookbooks/pacemaker/providers/primitive.rb (hunk 9 above)
Comment (AbelNavarro) [reply chef/cookbooks/pacemaker/providers/primitive.rb]: Tested that and works like a charm. Nice Ruby writing.

Comment (AbelNavarro) [reply chef/cookbooks/pacemaker/providers/primitive.rb]: Agreed, will move it to `set_op_default`.

File: chef/cookbooks/pacemaker/providers/primitive.rb (hunk 8 above)
Comment (AbelNavarro) [reply chef/cookbooks/pacemaker/providers/primitive.rb]: Agreed, will move it to `set_op_default` as well.

File: chef/cookbooks/pacemaker/providers/primitive.rb (hunk 6 above)
Comment (AbelNavarro) [reply chef/cookbooks/pacemaker/providers/primitive.rb]: Yes, it can be an array. This was introduced at https://github.com/crowbar/crowbar-ha/pull/282

File: chef/data_bags/crowbar/template-pacemaker.json (hunk 5 above)
Comment (AbelNavarro) [reply chef/data_bags/crowbar/template-pacemaker.json]: Right. I had that on the other PR but didn't port it here.

File: chef/cookbooks/pacemaker/providers/primitive.rb (hunk 7 above)
Comment (AbelNavarro) [reply chef/cookbooks/pacemaker/providers/primitive.rb]: Right, will try to find some nice Ruby syntax to make it clean.

**Comment by AbelNavarro** [comment]
CI keeps failing with `crowbarctl proposal commit database default` which is totally unrelated to this change. Rebuilding (again).

**Comment by AbelNavarro** [comment]
@vuntz green light. Time for another review please.

**Comment by AbelNavarro** [comment]
@aspiers mind having another look at this PR? danke!

**Comment by AbelNavarro** [comment]
@vuntz right, fixing that.

**Comment by AbelNavarro** [comment]
@vuntz @aspiers green light again. Mind having another look?

File: chef/cookbooks/pacemaker/providers/primitive.rb [hunk 10]
```diff
@@ -90,13 +96,49 @@ def update_resource(name)
   maybe_modify_resource(name)
 end
 
+def set_op_default(op_monitor, op_defaults)
```
Comment (aspiers) [inline chef/cookbooks/pacemaker/providers/primitive.rb]: I find this quite hard to understand, despite having been involved in its design.  I think a big part of the problem is in the variable naming:

- You have a method and a local variable both called `op_defaults`, which is really confusing.
- The variable names don't clearly distinguish between global `op` settings which comes from the Pacemaker proposal (via the node) and local `op` settings which come from the cookbook invoking `pacemaker_primitive`.  The whole point of this code is to take a copy of the global defaults and then merge in the per-cookbook settings without accidentally mutating any attributes along the way, but that's not obvious here.

In other words a lot of local knowledge is assumed, without any explanation.  Adding comments to explain is one possible solution, but that's often a bad smell.  If you can make the code clearer without adding (a lot of) comments, that would be much better.  So I suggest trying to rename the variables / methods to things like `global_op_defaults` and `local_op_defaults`, and adding a comment explaining why we are carefully avoiding mutation of data.  Then maybe it will be easier for others to follow.

Comment (AbelNavarro) [reply chef/cookbooks/pacemaker/providers/primitive.rb]: @aspiers thanks for the feedback. I understand it can be a problem to get the full context just reading the code. I made slight changes renaming vars in this commit: https://github.com/crowbar/crowbar-ha/pull/297/commits/4687e59ef3d9aa35ad0c6605060bfda15ff8ea59

I'm hesitant to do more changes because the code took long time to get it right. I'm open to more comments or even more vars/functions renaming though. How does it look like in this new approach?

File: chef/cookbooks/pacemaker/providers/primitive.rb [hunk 11]
```diff
@@ -69,6 +69,12 @@ def create_resource(name)
   standard_create_resource
 end
 
+def get_barclamp_op_defaults(node)
```
Comment (aspiers) [inline chef/cookbooks/pacemaker/providers/primitive.rb]: Thanks, this is definitely a step in the right direction, but the name is still vague because it looks like it could be referring to any barclamp (e.g. the one containing the currently running recipe), but in fact it refers specifically to the `pacemaker` barclamp.  That's why I previously suggested `global_op_defaults` to distinguish cluster-wide defaults from op values specific to a particular recipe/resource.  Having said that, it's per `pacemaker` proposal, which is actually per cluster rather than global like I originally thought.  So I think `get_cluster_op_defaults` would be better here.

File: chef/cookbooks/pacemaker/providers/primitive.rb [hunk 12]
```diff
@@ -90,13 +96,50 @@ def update_resource(name)
   maybe_modify_resource(name)
 end
 
+# Merge in op defaults from barclamp to those from pacemaker
+def set_op_default(op_monitor, op_defaults)
```
Comment (aspiers) [inline chef/cookbooks/pacemaker/providers/primitive.rb]: This method is actually not setting defaults, nor is it setting for all ops, but only for monitor.  So I suggest `merge_cluster_op_monitor_defaults`.

File: chef/cookbooks/pacemaker/providers/primitive.rb [hunk 13]
```diff
@@ -90,13 +96,50 @@ def update_resource(name)
...
+  ops = new_resource.op
+
+  # Convert to Hash the ops structure in case is in Chef::Node::Attribute form
+  # This is to avoid accidentally changing the original value
+  ops = ops.to_hash unless ops.is_a? Hash
+
+  ops ||= {}
+  ops["monitor"] ||= {}
+
+  barclamp_op_defaults = get_barclamp_op_defaults(node)
```
Comment (aspiers) [inline chef/cookbooks/pacemaker/providers/primitive.rb]: cluster_op_defaults = get_cluster_op_defaults(node)

**Comment by AbelNavarro** [comment]
Thanks @aspiers for the naming proposals. Those actually look much better than the ones I suggested. I think I've been too involved with this code to come up with with naming. I've added your suggestions in the last amended commit.

Btw, this extra commit to address the new naming with be squashed. For now it's much easier for me to have it separate.

**Comment by aspiers** [comment]
> I've added your suggestions in the last amended commit.

I don't see that :-/

> Btw, this extra commit to address the new naming with be squashed. For now it's much easier for me to have it separate.

Why is it easier?  It actually seems a lot harder to me, because it subverts GitHub's mechanism for marking comments as outdated.  Please just [squash and force-push every time](https://blog.adamspiers.org/2017/08/16/squash-merging-and-other-problems-with-github/).  If that is difficult for you then I suspect you are missing some simple git trick and I'd be happy to try to help make it easier for you.

**Comment by AbelNavarro** [comment]
@aspiers sorry, I forgot to push the changes yesterday, so should be no uploaded and squashed.

**Comment by AbelNavarro** [comment]
testbuild fails because the package from this PR (https://github.com/crowbar/crowbar-ha/pull/300) still isn't in the ISO. Will trigger a rebuild tomorrow.

File: chef/cookbooks/pacemaker/providers/primitive.rb [hunk 14]
```diff
@@ -90,13 +96,50 @@ def update_resource(name)
...
 
   Chef::Log.info "Checking existing #{@current_cib_object} for modifications"
 
   cmds = []
 
+  ops = new_resource.op
+
+  # Convert to Hash the ops structure in case is in Chef::Node::Attribute form
+  # This is to avoid accidentally changing the original value
+  ops = ops.to_hash unless ops.is_a? Hash
```
Comment (vuntz) [inline chef/cookbooks/pacemaker/providers/primitive.rb]: I think this is slightly wrong: if we were given a hash, we would modify it here, and the recipe would not expect that.

So I'd do something like:

```ruby
# We don't want to modify data from the recipe, so ensure we have a copy as a hash
if new_resource.op.nil?
  ops = {}
if new_resource.op.is_a? Hash
  ops = new_resource.op.clone
else
  ops = new_resource.op.to_hash
end
```

File: chef/cookbooks/pacemaker/providers/primitive.rb [hunk 15]
```diff
@@ -90,13 +96,50 @@ def update_resource(name)
...
+
+  ops ||= {}
```
Comment (vuntz) [inline chef/cookbooks/pacemaker/providers/primitive.rb]: That line could be dropped with my suggested code above.

File: chef/cookbooks/pacemaker/providers/primitive.rb (hunk 14 above)
Comment (aspiers) [reply chef/cookbooks/pacemaker/providers/primitive.rb]: Yep, that looks good to me.

Comment (aspiers) [reply chef/cookbooks/pacemaker/providers/primitive.rb]: > when you get it with new_resource.op you are getting a copy of it

Hrm, how do you know that?  And are you 100% sure it's a deep copy, rather than the values pointing to mutable data structures which are actually owned elsewhere?  If a caller passes in values from a cookbook, e.g.

```
  pacemaker_primitive ... do
    agent "..."
    op({
        start: { timeout: 120 },
        monitor: nova[:ha][:op][:monitor]
    })
    action :update
    only_if { CrowbarPacemakerHelper.is_cluster_founder?(node) }
  end
```

then I would expect `new_resource.op` to contain a `Hash` which has a `Chef::Node::Attribute` nested inside it, and that is not necessarily safe.

Admittedly I can't find any code where we actually do it like that now, but it's not safe to assume that it will never be written like that in the future.

Comment (AbelNavarro) [reply chef/cookbooks/pacemaker/providers/primitive.rb]: I can only talk from empiric experience: these hash values could not be modified. But let's agree to do that copy so we can be sure in other situations.

Notice though that the case you propose (hash containing Chef::Node::Attribute) is also error prone, because the hash.clone will just do a shallow copy on the Chef::Node::Attribute. This is the problem we originally had (I was doing Chef::Node::Atribute.clone and .dup) and the object was duplicated, but not the references. I hope we don't step into that case.

Comment (aspiers) [reply chef/cookbooks/pacemaker/providers/primitive.rb]: > I can only talk from empiric experience

Relying only on empirical evidence won't take care of all the cases :-)  You wouldn't have seen modifiable hash values because currently we don't nest `C::N::A` inside `Hash` anywhere.  We only construct hashes in places like https://github.com/crowbar/crowbar-openstack/blob/master/chef/cookbooks/postgresql/recipes/ha_storage.rb#L78 and in those cases they don't nest mutable objects.

> because the hash.clone will just do a shallow copy on the Chef::Node::Attribute

Sure, that's why I recommended a deep clone a few weeks / months ago when we were originally discussing how to handle this correctly.  [This StackOverflow article](https://stackoverflow.com/questions/8206523/how-to-create-a-deep-copy-of-an-object-in-ruby) might come in useful here.

**Comment by AbelNavarro** [comment]
@vuntz @aspiers your comments have been addressed. Build is green. Mind another review?

File: chef/cookbooks/pacemaker/providers/primitive.rb [hunk 16]
```diff
@@ -90,13 +96,51 @@ def update_resource(name)
...
 
+# Merge in op defaults from barclamp to those from pacemaker
+def merge_cluster_op_monitor_defaults(op_monitor, op_defaults)
+  return op_monitor if op_defaults.nil?
+
+  op_defaults ||= {}
+  op_defaults["monitor"] ||= {}
+  op_defaults["monitor"]["on-fail"] ||= ""
+
+  if !op_defaults["monitor"]["on-fail"].empty?
```
Comment (aspiers) [inline chef/cookbooks/pacemaker/providers/primitive.rb]: This seems needlessly complex.  Can't you just replace the whole of this method with these two lines?

    op_defaults ||= {}
    (op_defaults["monitor"] || {}).merge(op_monitor)

Comment (AbelNavarro) [reply chef/cookbooks/pacemaker/providers/primitive.rb]: The problem I see with this approach is that the merge is done at `op_defaults["monitor"]` level, while the original function did the merge at `op_defaults["monitor"]["on-fail"]` level. The intention for that was that the merge was doable for Hash and Array of Hashes. Although I agree the function is verbose I don't see any way to clean it up.

File: chef/cookbooks/pacemaker/providers/primitive.rb [hunk 17]
```diff
@@ -90,13 +96,51 @@ def update_resource(name)
...
+  elsif new_resource.op.is_a? Hash
+    ops = new_resource.op.clone
+  else
+    ops = new_resource.op.to_hash
+  end
+
+  cluster_op_defaults = get_cluster_op_defaults(node)
+
+  ops["monitor"] ||= {}
+  if ops["monitor"].is_a?(Array)
```
[the remark this answers is not in the document]
Comment (AbelNavarro) [reply chef/cookbooks/pacemaker/providers/primitive.rb]: Agree, adding a comment here.

File: chef/cookbooks/pacemaker/providers/primitive.rb [hunk 18]
```diff
@@ -90,13 +96,51 @@ def update_resource(name)
   maybe_modify_resource(name)
 end
 
+# Merge in op defaults from barclamp to those from pacemaker
+def merge_cluster_op_monitor_defaults(op_monitor, op_defaults)
+  return op_monitor if op_defaults.nil?
+
+  op_defaults ||= {}
```
[the remark this answers is not in the document]
Comment (AbelNavarro) [reply chef/cookbooks/pacemaker/providers/primitive.rb]: Agree, fixing that.

File: chef/cookbooks/pacemaker/providers/primitive.rb (hunk 16 above)
Comment (aspiers) [reply chef/cookbooks/pacemaker/providers/primitive.rb]: OK I think I see what you mean - here `op_defaults["monitor"]` will always be a `Hash` or `nil`, whereas `op_monitor` could be a `Hash` or an `Array`?  If so, again this is non-obvious and deserves a comment.

Comment (AbelNavarro) [reply chef/cookbooks/pacemaker/providers/primitive.rb]: Agree with that. I hope the added message will clarify things up.
